lib/ARSClient.mjs

```diff
@@ -272,7 +272,160 @@ class ARSClient {
     return retval;
   }
 
-  async _collectMergedResults(pkey, statusCheck = false) {
+  async _collectMergedResults(pkey, fetchCompleted=false) {
+    const ltype = fetchCompleted ? 'query-result' : 'query-status';
+
+    // Get the top level message from the ARS. This contains the list of currently
+    // merged PKs.
+    const [meta, arsResp] = await this._fetchMessage(pkey, true);
+    meta.timestamp = arsResp.timestamp || null;
+
+    /* Special case: if the ARS starts queuing requests, it will return status: running
+     * and an empty children array. Catch this special case and exit early.
+     */
+    if (arsResp.children.length === 0 && arsResp.status === "Running") {
+      logger.warn({ltype: ltype, pk: pkey}, 'Queuing')
+      return {
+        meta: meta,
+        pk: pkey,
+        queuing: true,
+        completed: [],
+        running: [],
+        errored: []
+      };
+    }
+
+    let mergedVersionsList = arsResp.merged_versions_list ?? null;
+    if (!mergedVersionsList || mergedVersionsList === "None") {
+      mergedVersionsList = [];
+    } else if (typeof mergedVersionsList === 'string') {
+      // Currently in the trace=y response, ARS incorrectly encodes MVL as a string instead of an actual array
+      mergedVersionsList = JSON.parse(mergedVersionsList.replace(/'/g, '"'));
+    }
+
+    let completed = [];
+    let running = [];
+    let errored = [];
+    let n_mv_incomplete = 0;
+    let n_mv_complete = 0;
+    let first_complete_mv_idx = null;
+    let fetch_completed_meta = null;
+
+    if (cmn.isArray(mergedVersionsList) && mergedVersionsList.length > 0) {
+      // Fetch all the merged version statuses to get the most recent version
+      // that is also complete.
+      mergedVersionsList.reverse(); // First element should be the most recent
+      const statusPromises = mergedVersionsList.map((mergedEntry) => {
+        const [uuid, agent] = mergedEntry;
+        return this._fetchMessage(uuid, true);
+      });
+
+      await Promise.allSettled(statusPromises).then((promises) => {
+        let i = 0;
+        for (const promise of promises) {
+          if (promise.status !== 'fulfilled') {
+            logger.warn(promise, `Unexpected rejection when fetching message`);
+            i += 1;
+            continue;
+          }
+          const message = promise.value[1]; // We do not care about status metadata
+          logger.debug(message, 'this is the promise value');
+          const status = {
+            agent: mergedVersionsList[i][1],
+            uuid: message.message,
+            status: message.status
+          };
+
+          // We have to inject the status codes because we expect them but the ARS does not
+          // provide them
+          if (message.status === 'Done' ||
+              (message.status === 'Error' && message.code === 444)) {
```

Comment:
I think we should add a logging message for the Error `444` case. This means that there was a minor error in annotation at some point upstream.